### PR TITLE
Linux: handle hugepages

### PR DESCRIPTION
**Pull request by cgzones**
Subtract hugepages from normal memory.
Add a HugePageMeter.

Closes: #447

/cc @davezarzycki

### CODE REVIEW

**Comment by davezarzycki** [comment]
Thanks for the super fast turnaround. This much better than my patch: #449 .

File: linux/LinuxProcessList.c [hunk 1]
```diff
@@ -1554,6 +1554,57 @@ static inline void LinuxProcessList_scanMemoryInfo(ProcessList* this) {
...
+      unsigned long int hugePageSize = strtoul(name + strlen("hugepages-"), &endptr, 10);
+      if (!endptr || *endptr != 'k')
+         continue;
+
+      char content[64];
+      char hugePagePath[128];
+      ssize_t r;
+
+      xSnprintf(hugePagePath, sizeof(hugePagePath), "/sys/kernel/mm/hugepages/%s/nr_hugepages", name);
+      r = xReadfile(hugePagePath, content, sizeof(content));
```
Comment (BenBE) [inline linux/LinuxProcessList.c:1586]: Maybe add `xReadfileat`?

File: linux/HugePageMeter.c [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      .display = HugePageMeter_display,
+   },
+   .updateValues = HugePageMeter_updateValues,
+   .defaultMode = BAR_METERMODE,
+   .maxItems = 1,
+   .total = 100.0,
+   .attributes = HugePageMeter_attributes,
+   .name = "HugePages",
+   .uiName = "HugePages",
+   .caption = "HuP"
```
Comment (BenBE) [inline linux/HugePageMeter.c:62]: Just `HP ` maybe?

File: linux/LinuxProcessList.c (hunk 1 above)
Comment (cgzones) [reply linux/LinuxProcessList.c:1586]: For `xReadfileat()` we need a file-descriptor of the parent directory; so we'd have to open `/sys/kernel/mm/hugepages/`, then open the distinct pagesize directory (e.g. `hugepages-2048kb`) and then pass that to `xReadfileat()`.

As `LinuxProcessList_scanHugePages()` is only called once every update cycle I avoided that.

File: linux/HugePageMeter.c (hunk 2 above)
Comment (davezarzycki) [reply linux/HugePageMeter.c:62]: For whatever it may be worth, I still think that people that use huge pages will want a meter per page size. So for example, x86-64 supports 2 MiB and 1 GiB huge pages. On other architectures, the available huge page sizes aren't friendly to two or three character abbreviations. For example, on arm64, a 512 MiB huge page size is possible: https://www.kernel.org/doc/html/latest/arm64/hugetlbpage.html

Comment (cgzones) [reply linux/HugePageMeter.c:62]: Sadly Meters do not (yet) support options, so currently we would have to create (at compile time) Meters for all possible/supported page sizes.
As long we support *only* 2, 512 and 1024 MiB that might be a workaround.

Comment (davezarzycki) [reply linux/HugePageMeter.c:62]: I was thinking about this some more. We could just have different values for "the" huge page meter, so rather than "used" versus "total", it becomes "used-2m", "used-1g", "total". And the most number of huge page sizes I've seen supported at the same time seems to be four. What do you think?

Comment (BenBE) [reply linux/HugePageMeter.c:62]: We already support for meters to show more than just one value. So this probably won't be the main issue to show them. What might be somewhat harder, when not falling back to hardcoding the actual sizes is to transmit the actual labels. If the size of hugepages are more or less known, you could also just alloc an array with say 24 entries (1MiB to 8TiB) and just put the proper values in the right slot or `NAN` if unsupported. On the Meter side, coloring would than just count the used slots with the first used getting color 1, the second color 2, and so on. Sure some more LOC, but wouldn't limit extensibility too much and avoids hardcoding the labels (HugePages that ain't Power of Two make little sense).

Comment (davezarzycki) [reply linux/HugePageMeter.c:62]: So think I've more or less implemented this model. Please see the following commit that builds on this pull request: https://github.com/davezarzycki/htop/commit/410c24341deaceb1b539fc8cad3873b0fb813eef

Comment (davezarzycki) [reply linux/HugePageMeter.c:62]: Thanks for the feedback. Here is my updated branch that is derived from this pull request: https://github.com/davezarzycki/htop/tree/per_hugepage_values

Comment (davezarzycki) [reply linux/HugePageMeter.c:62]: The above branch now has a fix for the hasty `nextUsed` change I just made.

Comment (davezarzycki) [reply linux/HugePageMeter.c:62]: I think I've addressed all of the feedback to date, including the import order change that I overlooked yesterday.

I've also updated the branch after two additional tests where zero or one huge page size is available (rather than the two sizes on my system). Now zero or one available huge page size will print " used:" instead of the page size, and zero available huge page sizes are now treated as if one huge page size exists (with zero used and zero total).

**Review by BenBE — CHANGES_REQUESTED** [review]

**Comment by davezarzycki** [comment]
@cgzones – I've rebased and updated my `per_hugepage_values` branch based on this pull request. I think I've addressed all but one the feedback notes @BenBE has provided. I don't feel any emotional attachment to my patch that builds on your pull request, so if you want to just "squash" it into your pull request, that's okay with me. If credit matters to you, then I can just keep updating my branch based on your pull request.

@BenBE – I'm not sure what you mean by [this comment](https://github.com/htop-dev/htop/pull/450#discussion_r558847538). That `total` field is effectively `const`, is it not? In any case, the update function does update the total and I haven't figured out a way yet where that can go awry.

**Review by BenBE — APPROVED** [review]

**Comment by cgzones** [comment]
Thanks @davezarzycki for contribution.
